### DISCUSSION TITLE
Applying migrations in language bindings

### DISCUSSION THREAD

**Discussion by tailhook** [geldata/gel#4758] [Pre-RFC]
# Motivation

Users want to make their applications standalone. This includes not relying on `edgedb` command-line tool for running applications.

Creating, modifying, schema diffing and so other operations related to migrations are quite complex ones. We want to improve developer experience with these without the need of updating all the language bindings. On the other hand, applying already ready and validated migrations is relatively stable operation that can be done by language bindings.

# Requirements

1. We want to avoid the need for having a parser in language bindings (currently parser is needed to get migration ID from the file)
2. We need to allow feeding migrations from sources other than the filesystem (examples are: embedding migrations in the binary itself for compiled languages, reading from `.zip` in scripting languages).
3. Migration source comes with a list of migrations that is ordered apropriately.

# API

## General Strategy

Connection pool:
1. `apply_migrations` method that accepts argument of `MigrationSource` interface.
2. All migrations are applied within the single transaction.
3. If at any migration transaction is failed due to `DuplicateMigrationError` (see edgedb/edgedb#4747), it closed. New transaction is opened and process continues starting from the next migration. In all other senses, the error is ignored.

Applying within user's transaction:
1. `Transaction` object also has `apply_migrations`.
2. Each migration is applied in their own savepoint.
3. Queries failed by `DuplicateMigrationError` rollback to savepoint, and error is ignored.
4. This method allows adding application-specific stuff into the transaction doing migration.
5. We don't add any hooks for per-migration processing. But if it's needed it can be crafted by using a loop and a migration source per transaction.

## Rust

```
trait MigrationSource {
    async fn set_last_applied(mut self, name: &str) {
        // default implementation does nothing
    }
    async fn revisions(&mut self) -> Box<Stream<&str>+'_>;
}
impl Client {
    async fn apply_migrations(&self, src: impl MigrationSource)
        -> Result<(), edgedb_tokio::Error>;
}
impl Transaction {
    async fn apply_migrations(&mut self, src: impl MigrationSource)
        -> Result<(), edgedb_tokio::Error>;
}

struct FsMigrationSource { ... };
impl MigrationSource for FsMigrationSource {
    /// Assumes `dbschema/migrations` dir
    fn open_dir(dir: &Path) -> Result<(), edgedb_tokio::Error>; 
}
```

Note: default implementation of migration source from the filesystem will read all the revisions and rely on the `apply_migrations` to skip already applied ones. Other sources can possibly optimize by starting from the revision next to `set_last_applied`.

## Python

```
class AsyncIOMigrationSource(abc.ABC):

    async def set_last_applied(self, name: str) -> None:
        # default implementation does nothing
        pass

    @abc.abstracmethod
    async def revisions(self) -> AsyncIterable[str]: ...

class Client:
    # ...
    async def apply_migrations(self, source: AsyncIOMigrationSource) -> None: ...

class AsyncIOIteration:   # this is the actual transaction class
    # ...
    async def apply_migrations(self, source: AsyncIOMigrationSource) -> None: ...

class FSMigrationSource:

    def __init__(self):
        """assumes normal project resolution and reading `edgedb.toml`"""
        if not isPy_3_11:
            throw NotImplementedError("no tomllib in python < 3.11")
        # ...

    @classmethod
    def open_dir(cls, project_dir: pathlib.Path):
        """assumes `dbschema/migrations` dir, does not need tomllib"""
        ...
```

Note: default implementation of migration source from the filesystem will read all the revisions and rely on the `apply_migrations` to skip already applied ones. Other sources can possibly optimize by starting from the revision next to `set_last_applied`.

# Open Questions

1. Python specific: should `apply_migrations` accept `AsyncIOMigrationSource | MigrationSource` to simplify user implementations? (as most sources will probably be synchronous ones: either filesystem or memory)

**Comment by elprans** [comment]
> Python specific: should `apply_migrations` accept `AsyncIOMigrationSource | MigrationSource` to simplify user implementations?

I'd say yes.  Supporting various iterator inputs isn't hard.

**Comment by elprans** [comment]
> `class FSMigrationSource:`

I don't think a complex interface is really necessary here, we should be able to simply have `apply_migrations` accept an iterable of strings directly and return the name of the last applied migration.

**Reply by tailhook** [reply]
There are few things to consider:
1. I think optimizing starting revision might make sense in some cases. One example is probably fetching revisions from a remote source.
2. This is also good for documentation: you can look up things that implement specific trait/interface.
3. For some languages like python, where any iterator is iterable, people might use some generator there. And it will work, until transaction needs to be retried.

The alternative for the (1) is make a method `pool.get_last_migration()`. So users can do that themselves. But this makes harder for them to switch between the sources I think (i.e. one for local and other for production).

**Reply by elprans** [reply]
Sorry, I still don't follow here.  To me the `AsyncIOMigrationSource` pattern presented above is not more functional than a simple `def apply_migrations(AsyncIterable[str]) -> str`.  What am I missing?

> For some languages like python, where any iterator is iterable, people might use some generator there. And it will work, until transaction needs to be retried.

But `AsyncIOMigrationSource.revisions` also returns an iterable above?  If that's considered a problem, then we can mandate a materialized list as input.  Migrations aren't supposed to be very large anyway, so supporting streaming is probably not necessary.

**Reply by tailhook** [reply]
> Sorry, I still don't follow here. To me the `AsyncIOMigrationSource` pattern presented above is not more functional than a simple `def apply_migrations(AsyncIterable[str]) -> str`. What am I missing?
> 

To optimize which revisions to fetch, `MigrationSource` receives a starting revision via `set_last_applied`. And then `.revisions()` can return revisions starting from the next one.

We will not use it for default `FSMigrationSource` because that requires parsing, but can use that for migrations embedded into the binary (i.e. parse them at compile time and keep an ordered map of migrations). So applying migrations is a little bit faster.

And as I've written before if someone decides to fetch migrations from remote source this might be useful too.

Also Rust docs display implementors for the trait:
<img src="https://user-images.githubusercontent.com/321799/205939025-4a6b7a64-bc84-46a3-8b8a-026b89292ad2.png" width="361" />
So it's easier to find what you can specify as an argument to `apply_migrations`, rather than looking for a function that returns `IntoIter<Item=AsRef<str>>` or `Vec<String>` (which there will be a lot of ones).

I'm not sure if Python/Sphinx have this feature. But I think modern IDEs are happy to suggest things based on type.

> > For some languages like python, where any iterator is iterable, people might use some generator there. And it will work, until transaction needs to be retried.
> 
> But `AsyncIOMigrationSource.revisions` also returns an iterable above?

Yes. But can be called multiple times.

> If that's considered a problem, then we can mandate a materialized list as input. Migrations aren't supposed to be very large anyway, so supporting streaming is probably not necessary.

I don't think making this unoptimizable ever is a great idea. Concerns are:

1. Memory usage on small devices/servers
2. Performance (i.e. skipping 1000 of the revisions might be an issue).

**Reply by elprans** [reply]
OK, I see now.  Might still be still worth _also_ supporting a simple iterator (at least in Python).
